### DISCUSSION TITLE
Remote Debugging Quarkus

### DISCUSSION THREAD

**Discussion by s3nu4** [keycloak/keycloak#12679] [Archive]
Hey,

I updated my Keycloak installation from 16 (WildFly) to 18 (Quarkus). Until the update I was able to debug my custom SPI extensions with remote debugging in IntelliJ. I just had to add the DEBUG: "true" and DEBUG_PORT: "*:8787" to my docker-compose setup. 

I know that the framework has changed to Quarkus and the way to remote debug  has probably changed, but I can't find out how to do this with Keycloak 18. 

Do you have any advices how to enable remote debugging? 

Thank you!

**Comment by MetroMarv** [comment]
Actually I was also not able to find any documentation on this. Also `kc.sh --help` does not list any configuration options on debugging.

However this article https://blog.codecentric.de/en/2021/12/keycloak-keycloak-x/ says the you need to use the `--debug` flag when starting the Keycloak. And additionally specify the env var `DEBUG_PORT='*:8787'`. This also worked for me.

**Reply by MetroMarv** [reply]
Actually the [quarkus docs](https://access.redhat.com/documentation/en-us/red_hat_build_of_quarkus/1.3/html/developing_and_compiling_your_quarkus_applications_with_apache_maven/proc-debugging-your-quarkus-project_quarkus-maven) give some more insight on debugging

**Comment by martin-kanis** [comment]
@DGuhr @pedroigor Guys, I think we should add information about Quarkus remote debugging into the help and to the docs if is not there? WDYT?

**Reply by DGuhr** [reply]
Sure. We haven't started writing developer guides yet, always fell down the road due to other priorities. Feel free to create an issue :)

For other people landing here, to remote debug, do the following: 

start Keycloak using `kc.<sh|bat> --debug <start|build|start-dev> ...`
a remote debugging session will be initiated using port 8787 by default. 

Some options:
- to suspend your debugging session, you can set an environment variable -> `export DEBUG_SUSPEND=y`
- to set the port, use the env var -> `export DEBUG_PORT=<port>`
  - to be able to remote debug a containerized keycloak, set the port to `export DEBUG_PORT=*:<port>`

When a build is invoked as part of the server startup (so not using --optimized and having some changes that trigger a rebuild), another JVM is spawned internally, so your debug session will be killed when the build is done and we're back in the "runtime" vm. make sure to reconnect after the `build` process is done. Best to do with the suspend flag ;)

**Reply by SangI762** [reply]
Thanks, it worked.

**Comment by martin-kanis** [comment]
Here is the issue https://github.com/keycloak/keycloak/issues/13513
Thank you for explaining the fact, why I need to connect my debugger twice :)

**Comment by jsorah** [comment]
Throwing another way to get this with a containerized run using podman (should work for docker too):

```
podman run -p 8080:8080 -p 8443:8443 -p <port>:<port> -e DEBUG_PORT='*:<port>' -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin quay.io/keycloak/keycloak:19.0.1 --debug start-dev
```

and then attach your debugger to `<port>` where `<port>` is your desired debug port.

**Reply by lbras** [reply]
Thanks, the `-e DEBUG_PORT='*:<port>'` was crucial for getting the debugger to attach using docker.

**Reply by StefanLobbenmeierObjego** [reply]
For me on Keycloak 26 (quay.io/keycloak/keycloak:26.4.0) it was 

```docker-compose.yml
    env:
      DEBUG_PORT: '30012'
    command:
      - "--debug"
      - "start-dev"
      - "--import-realm"
```


(no source IP needed)

**Comment by EvgeniGordeev** [comment]
here is an [example Maven project](https://github.com/EvgeniGordeev/keycloak-quarkus-debug) to debug Quarkus KC easily in an IDE. 

NB: Using production `start` instead of `start-dev` helps to avoid connecting the debugger twice.
